// gl-engine/es5/lib/material/flat/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createFlatMaterial;

var _setUniforms = require('./set-uniforms');

var _setUniforms2 = _interopRequireDefault(_setUniforms);

var _setAttributes = require('./set-attributes');

var _setAttributes2 = _interopRequireDefault(_setAttributes);

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

var _useAugment = require('../../augment/use-augment');

var _useAugment2 = _interopRequireDefault(_useAugment);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Browserify Transforms:

function FlatMaterial(shadingConfig) {
  this.flags = {
    visible: true,
    transparent: false,
    recompile: true,
    useLights: false
  };

  this.defines = ['CAMERA'];
  this.shader = null;
  this.mode = 'TRIANGLES';
  this.setUniforms = [_setUniforms2.default];
  this.setAttributes = [_setAttributes2.default];
  this.shading = shadingConfig;
  this.use = _useAugment2.default.bind(null, this);
}

FlatMaterial.prototype.fragSource = require('fs').readFileSync(__dirname + '/flat.built.frag', 'utf8');
FlatMaterial.prototype.vertSource = require('fs').readFileSync(__dirname + '/flat.built.vert', 'utf8');

function createFlatMaterial(properties) {
  var shadingConfig = (0, _objectAssign2.default)({
    color: [0.1, 0.1, 0.1],
    opacity: 1.0
  }, properties);

  return new FlatMaterial(shadingConfig);
}
