// gl-engine/es5/lib/renderer/multipass/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createMultiPass;

var _usePass = require('./use-pass');

var _usePass2 = _interopRequireDefault(_usePass);

var _render = require('./render');

var _render2 = _interopRequireDefault(_render);

var _destroy = require('./destroy');

var _destroy2 = _interopRequireDefault(_destroy);

var _glFbo = require('gl-fbo');

var _glFbo2 = _interopRequireDefault(_glFbo);

var _debug = require('./debug');

var _debug2 = _interopRequireDefault(_debug);

var _getDepth = require('./get-depth');

var _getDepth2 = _interopRequireDefault(_getDepth);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function MultiPass(renderer, createDebugger) {
  this.renderer = renderer;
  this.gl = renderer.gl;
  this.passes = [];
  this.resolution = [renderer.canvas.width, renderer.canvas.height];
  this.fbos = [(0, _glFbo2.default)(renderer.gl, this.resolution), (0, _glFbo2.default)(renderer.gl, this.resolution)];

  this.fbos[0].name = 'fbo_0';
  this.fbos[1].name = 'fbo_1';

  this.use = _usePass2.default.bind(null, this);
  this.render = (0, _render2.default)(this);
  this.destroy = _destroy2.default.bind(null, this);
  this.debug = createDebugger ? createDebugger(this.gl) : null;
  this.canReadDepth = this.gl.getExtension('WEBGL_depth_texture');
  // this.canReadDepth = false
  this.getDepth = (0, _getDepth2.default)(this);
}

function createMultiPass(forwardRenderer, createDebugger) {
  if (!forwardRenderer) {
    throw new Error('Multipass renderer must be provided a forward renderer.');
  }
  return new MultiPass(forwardRenderer, createDebugger);
}
