// gl-engine/es5/tests/test-pixel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createPixelTester;

var _readPixel = require('./read-pixel');

var _readPixel2 = _interopRequireDefault(_readPixel);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Fuzzily checks a pixel
function testPixel(t, a, b, message, epsilon) {

  var pass = a.length === b.length;
  for (var i = 0; i < b.length; i++) {
    pass = pass && Math.abs(a[i] - b[i]) < epsilon;
  }
  t._assert(pass, {
    message: message,
    operator: 'testPixel',
    actual: a,
    expected: b
  });
}

function createPixelTester(t, gl) {
  var epsilon = arguments.length <= 2 || arguments[2] === undefined ? 2 : arguments[2];

  return function runPixelTester(x, y, comparison) {
    var message = arguments.length <= 3 || arguments[3] === undefined ? 'pixels do not match' : arguments[3];

    var pixel = (0, _readPixel2.default)(gl, x, y);
    testPixel(t, pixel, comparison, message, epsilon);
  };
}
